velox/functions/prestosql/DecimalFunctions.cpp

```diff
@@ -197,7 +197,10 @@ struct DecimalDivideFunction {
 
   inline static uint8_t
   computeRescaleFactor(uint8_t fromScale, uint8_t toScale, uint8_t rScale) {
-    return rScale - fromScale + toScale;
+    auto rescaleFactor = rScale - fromScale + toScale;
```

Comment:
It took me some time to understand how / why we compute rescaleFactor this way, but I think I get it.

I also noticed that the calling code is more complex then necessary. It first computes result precision and scale, then computes rescale factor using only result scale:

```
auto [rPrecision, rScale] =
        computeResultPrecisionScale(aPrecision, aScale, bPrecision, bScale);
    aRescale_ = computeRescaleFactor(aScale, bScale, rScale);
```

It would be nice to simply:

```
const auto rScale = std::max(aScale, bScale);
aRescale_ = rScale - aScale + bScale;
VELOX_USER_CHECK_LE(aRescale_, ...)
```

Also, should we use USER_CHECK here to generate user error and not system error?

It is also a bit sad that this requirement cannot be expressed as part of the signature to ensure that the error happens during query compilation rather than during query execution.

Let's document this restriction in https://facebookincubator.github.io/velox/functions/presto/decimal.html#divide

Maybe  we could say something like

```
Throw if max(s1, s2) - s1 + s2 exceeds 38.
```

It might be helpful to also extend documentation in https://facebookincubator.github.io/velox/functions/presto/decimal.html#division to explain how division is calculated and where the limitation above is coming from.

Reply:
Maybe we could add something like this to the docs: https://facebookincubator.github.io/velox/functions/presto/decimal.html#division 

```
Let’s take a closer look at how division is implemented.

Let’s say a is a decimal(p1, s1) with unscaled value A and b is a decimal(p2, s2) with unscaled value B. 

a = A / 10^s1
b = B / 10^s2

a / b = A * 10^(s2 - s1) / B

We decided that the scale of the result is s = max(s1, s2). Hence, unscaled value of the result would be

A * 10^(s + s2 - s1) / B

To compute this value, we first rescale A using the rescale factor (s + s2 - s1), then divide by B and round to the nearest whole. This method works as long as the rescale factor doesn’t exceed 38. If s + s2 - s1 exceeds 38 an error is raised.
```

I also notice that the paragraph starting with "The choice of formula for the..." doesn't make sense. Perhaps, we can remove it.

Reply:
I see that doc is already taken care of here: https://github.com/facebookincubator/velox/pull/9380

Reply:
> I also notice that the paragraph starting with "The choice of formula for the..." doesn't make sense. Perhaps, we can remove it.

@karteekmurthys https://github.com/facebookincubator/velox/pull/9380 only took care of this one thing.

Reply:
Updated the docs.